server/interceptor/interceptor_test.go

```diff
@@ -0,0 +1,310 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interceptor
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/storage/testonly"
+	"github.com/google/trillian/trees"
+	"github.com/kylelemons/godebug/pretty"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+func TestTreeInterceptor_UnaryInterceptor(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	logTree := *testonly.LogTree
+	logTree.TreeId = 10
+	mapTree := *testonly.MapTree
+	mapTree.TreeId = 11
+	unknownTreeID := int64(999)
+
+	admin := storage.NewMockAdminStorage(ctrl)
+	adminTX := storage.NewMockReadOnlyAdminTX(ctrl)
+	admin.EXPECT().Snapshot(gomock.Any()).AnyTimes().Return(adminTX, nil)
+	adminTX.EXPECT().GetTree(gomock.Any(), logTree.TreeId).AnyTimes().Return(&logTree, nil)
+	adminTX.EXPECT().GetTree(gomock.Any(), mapTree.TreeId).AnyTimes().Return(&mapTree, nil)
+	adminTX.EXPECT().GetTree(gomock.Any(), unknownTreeID).AnyTimes().Return(nil, errors.New("not found"))
+	adminTX.EXPECT().Close().AnyTimes().Return(nil)
+	adminTX.EXPECT().Commit().AnyTimes().Return(nil)
+
+	tests := []struct {
+		desc       string
+		req        interface{}
+		handlerErr error
+		wantErr    bool
+		wantTree   *trillian.Tree
+	}{
+		{
+			desc: "rpcWithoutTree",
+			req:  &trillian.CreateTreeRequest{},
+		},
+		{
+			desc:     "adminRPC",
+			req:      &trillian.GetTreeRequest{TreeId: logTree.TreeId},
+			wantTree: &logTree,
+		},
+		{
+			desc:     "logRPC",
+			req:      &trillian.GetLatestSignedLogRootRequest{LogId: logTree.TreeId},
+			wantTree: &logTree,
+		},
+		{
+			desc:     "mapRPC",
+			req:      &trillian.GetSignedMapRootRequest{MapId: mapTree.TreeId},
+			wantTree: &mapTree,
+		},
+		{
+			desc:    "unknownRequest",
+			req:     "not-a-request",
+			wantErr: true,
+		},
+		{
+			desc:    "unknownTree",
+			req:     &trillian.GetTreeRequest{TreeId: unknownTreeID},
+			wantErr: true,
+		},
+	}
+
+	ctx := context.Background()
+	intercept := TreeInterceptor{Admin: admin}
+	for _, test := range tests {
+		handler := &fakeHandler{resp: "handler response", err: test.handlerErr}
+
+		resp, err := intercept.UnaryInterceptor(ctx, test.req, &grpc.UnaryServerInfo{}, handler.run)
+		if hasErr := err != nil && err != test.handlerErr; hasErr != test.wantErr {
+			t.Errorf("%v: UnaryInterceptor() returned err = %q, wantErr = %v", test.desc, err, test.wantErr)
```

Comment:
I still prefer it. It does look a bit strange on nils, but I find it to be much nicer on non-nils.